### ISSUE TITLE
Media worker accumulates over 1000 connections to mega.nz

### ISSUE THREAD

**Issue by syldrathecat** [matrix-org/synapse#15440]
### Description

Our homeserver is accumulating connections to mega.nz (31.216.145.5) at a rate of approximately 2 per hour, most likely corresponding to URL preview requests for mega.nz. The largest amount I've seen is 1025 connections.

### Steps to reproduce

- Request a moderate amount of mega.nz url previews over a long period of time
- Observe the effect on the server via e.g. `netstat -an | grep 31.216.145.5`

I am observing the same https connections being open from 10 hours ago, as well as additional new ones opening.

The URL preview requests complete as expected, but the https connections are never closed, and new ones are slowly added over time.

I would expect that at some point the connections would be closed, rather than being staying open indefinitely.

### Homeserver

another homeserver

### Synapse Version

1.81

### Installation Method

Debian packages from packages.matrix.org

### Database

PostgreSQL

### Workers

Multiple workers

### Platform

lxc container with no outgoing http proxy configured for url previews

### Relevant log output

```shell
-
```

**Comment by reivilibre** [comment]
Do you have an example mega.nz URL that causes this, for us to debug with? :)

**Comment by syldrathecat** [comment]
> Do you have an example mega.nz URL that causes this, for us to debug with? :)

https://mega.nz/folder/test1
https://mega.nz/folder/test2

etc. will trigger URL preview requests

Testing it on a separate server, previewing those URLs also opens connections that seem to never close -- while another url such as https://example.org/ will close after a few minutes.

**Comment by DMRobertson** [comment]
Possibly related: https://github.com/matrix-org/synapse/issues/8302, https://github.com/matrix-org/synapse/security/advisories/GHSA-4822-jvwx-w47h

**Comment by DMRobertson** [comment]
> Synapse doesn't have any special handling of keep-alive that I can see; it delegates to `treq` to make the HTTP GET call

We don't pass a timeout directly to `treq`; instead we use our own timeout mechanism:

https://github.com/matrix-org/synapse/blob/c9326140dc9f8ea849356b5a8397e468636df8d4/synapse/http/client.py#L406-L412

But this internally calls `cancel()` on the deferred from `treq` here

https://github.com/matrix-org/synapse/blob/c9326140dc9f8ea849356b5a8397e468636df8d4/synapse/http/client.py#L406-L412

(which has the same effect as passing a timeout to treq.request).

We also have a timeout for reading the body of the request: https://github.com/matrix-org/synapse/blob/c9326140dc9f8ea849356b5a8397e468636df8d4/synapse/http/client.py#L728

So TL;DR I can't see an obvious path in Synapse that would mean url preview connections hang around longer than they should do.

----

Are there any relevant logs including the string `mega.nz` on your media workers?

**Comment by clokep** [comment]
I looked into this a little bit and found some interesting (?) things:

* Synapse [creates a custom HTTP connection pool](https://github.com/matrix-org/synapse/blob/c9326140dc9f8ea849356b5a8397e468636df8d4/synapse/http/client.py#L792-L800) with (by default) up-to 50 connections per host which time out after 2 minutes.
* After waiting 2 minutes the connection is gone from the pool, but using `netstat` I'm still seeing it.

The default number of connections per host that Twisted usually has is 2, I think this is way overkill from the URL previewer. The comments say the number was upped for connections to push servers; I think we should have a separate `SimpleHttpClient` for that then.

**Comment by clokep** [comment]
From some conversations with the Twisted developers, it seems like this is related to TLS not shutting down properly. (There's a bunch of issues about this on the Twisted bug tracker.) Will try to investigate a bit more.

**Comment by clokep** [comment]
I think twisted/twisted#7926 is the canonical upstream issue, the tl;dr is that you sometimes end up in OpenSSL calling [`SSL_shutdown`](https://www.openssl.org/docs/manmaster/man3/SSL_shutdown.html) which returns `False` meaning the TLS connection hasn't been shutdown yet. Twisted does nothing (?) in this situation and the underlying TCP connection is kept open. That's my understanding at least.

I'm not quite sure *what* the proper behavior here is though, it sounds like you should attempt to read (and write?) to the TLS connection until a response is received?

Relevant twisted code is at https://github.com/twisted/twisted/blob/7e57b24a9fb20c590eac97d7b8ba0730c3633873/src/twisted/protocols/tls.py#L331-L349

One of the developers suggested some of the hacks in https://gist.github.com/adiroiban/0824467069b25f308b1c71249b146132 might be useful, but I haven't tried them.

**Comment by clokep** [comment]
I don't have time to work on this at the moment but to re-iterate I think there's two things to follow up on here:

1. Limit the size of the HTTP connection pool used for URL previews. This would mean re-landing #15514. (It was backed out in #15530 due to regression).
2. Figure out how to fix the TLS connection leak inside of Twisted.
